hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-core/src/main/java/org/apache/hadoop/mapred/IFile.java

```diff
@@ -433,8 +434,11 @@ public boolean nextRawKey(DataInputBuffer key) throws IOException {
     }
     
     public void nextRawValue(DataInputBuffer value) throws IOException {
+      long targetSizeLong = currentValueLength + (long) (currentValueLength >> 1);
+      int targetSize = (int) Math.min(targetSizeLong, ARRAY_MAX_SIZE);
+
       final byte[] valBytes = (value.getData().length < currentValueLength)
-        ? new byte[currentValueLength << 1]
+        ? new byte[targetSize]
```

Comment:
just to be safe, let's keep the doubling but set a max value (Integer.MAX_VALUE - 8)